sdcm/cluster_aws.py

```diff
@@ -370,6 +372,36 @@ def destroy(self):
         self._instance.terminate()
         self.log.info('Destroyed')
 
+    def stop_task_threads(self, timeout=10):
+        if self._spot_aws_termination_task and not self.termination_event.isSet():
+            self.termination_event.set()
+            self._spot_aws_termination_task.join(timeout)
+        super(AWSNode, self).stop_task_threads(timeout)
+
+    def get_aws_termination_notification(self):
+        try:
+            result = self.remoter.run('curl http://169.254.169.254/latest/meta-data/spot/instance-action', verbose=False)
+            status = result.stdout.strip()
+            if '404 - Not Found' not in status:
+                return status
+        except Exception as details:
+            self.log.error('Error during getting aws termination notification %s' % details)
+        return None
+
+    def monitor_aws_termination_thread(self):
+        while True:
+            if self.termination_event.isSet():
+                break
+            self.wait_ssh_up(verbose=False)
+            result = self.get_aws_termination_notification()
+            if result:
+                self.log.warning('Got spot termination notification from AWS %s' % result)
+            time.sleep(5)
```

Comment:
Why do you use fixed 5 seconds? I suggested to record the time at the beginning of while loop, and sleep reset time.
```
        while True:
            start = time.time()
            if self.termination_event.isSet():
                break
            self.wait_ssh_up(verbose=False)
            result = self.get_aws_termination_notification()
            if result:
                self.log.warning('Got spot termination notification from AWS %s' % result)
            time.sleep(5 - (time.time() - start))
```

Reply:
Amazon recommend check status each 5 seconds, therefore used sleep for specific period.
Can't get idea of your code. How will help to set pause. Also, time.sleep can't get the negative value.

Reply:
sleep shell command can acceptable negative value, treat it as 0.
```
$ time sleep -10

real	0m0.004s
user	0m0.001s
sys	0m0.003s
```

What I want to improve is the real check interval.
Let's call the time that wait_ssh_up(), get_aws_termination_notification() take as `Execute time`.

If the `Execute time` is too big, then the real check interval will be increased, it equals to `Execute time` + sleep time (5 seconds).

My code tried to get reset sleep time by reducing the `Execute time` from 5 second. If `Execute time` <= 5 seconds, then the real check interval is 5 seconds.

 ```
        while True:
            start = time.time()
            if self.termination_event.isSet():
                break
            self.wait_ssh_up(verbose=False)
            result = self.get_aws_termination_notification()
            if result:
                self.log.warning('Got spot termination notification from AWS %s' % result)
            rest_time = 5 - (time.time() - start)
            if rest_time > 0:
                time.sleep(rest_time)
```

It's not a serious problem, it's fine to use your code.

Reply:
@amoskong a good point when we need to sleep near or **exactly** 5 seconds.
To improve the predictability of iteration time, we can move the `self.wait_ssh_up` out from the while loop.
@aleksbykov is this possible?